File: gtest/test_addbmm.cpp
#include <gtest/gtest.h>
#include "addbmm.h"  // Include the header for your custom addbmm class

// Define a struct for test parameters
struct AddBMMTestParams {
    torch::Tensor input;
    torch::Tensor batch1;
    torch::Tensor batch2;
    double beta;
    double alpha;
};

// Define the CustomAddBMMTest class
class CustomAddBMMTest : public ::testing::TestWithParam<AddBMMTestParams> {
protected:
    custom_namespace::AddBMM* custom_addbmm;

    // Setup function to initialize the test fixture
    void SetUp() override {
        custom_addbmm = new custom_namespace::AddBMM();
    }

    // Tear down function to clean up after tests
    void TearDown() override {
        delete custom_addbmm;
    }
};

// Instantiate parameterized tests with different input parameters
INSTANTIATE_TEST_SUITE_P(
    CustomAddBMMTests,
    CustomAddBMMTest,
    ::testing::Values(

        // Random tensor with uniform distribution
        AddBMMTestParams{torch::rand({2, 3}), torch::rand({4, 2, 3}), torch::rand({4, 3, 3}), 1.0, 1.0},

        // Random tensor with uniform distribution
        AddBMMTestParams{torch::rand({2, 3}), torch::rand({4, 2, 8}), torch::rand({4, 8, 3}), 1.0, 1.0},

        // Random tensor with uniform distribution with larger batch size for mat1 and mat2 
        AddBMMTestParams{torch::rand({3, 6}), torch::rand({1, 3, 1}), torch::rand({1, 1, 6}), 2.0, 1.0}

    )
);

// Parameterized test for Custom AddBMM function
TEST_P(CustomAddBMMTest, ParameterizedAddBMM) {
    AddBMMTestParams params = GetParam();

    torch::Tensor result = custom_addbmm->compute_addbmm(params.input, params.batch1, params.batch2, params.beta, params.alpha);
    
    // Expected result: manually perform the matrix multiplication and addition
    torch::Tensor expected = torch::addbmm(params.input, params.batch1, params.batch2, params.beta, params.alpha);

    // Compare results with a tolerance of 1e2
    ASSERT_TRUE(torch::allclose(result, expected, 1e2));
}

File: test/test_addbmm.py
import torch, pytest
import custom_module

torch.manual_seed(2)


@pytest.mark.parametrize(
    "input, batch1, batch2",
    (
        (
            torch.randn(3, 2),
            torch.randn(4, 3, 3),
            torch.randn(4, 3, 2),
        ),  # Regular input and matrices
        (
            torch.zeros(2, 3),
            torch.ones(4, 2, 3),
            torch.ones(4, 3, 3),
        ),  # Zeros input, ones for mat1, mat2
        # Varying tensor shapes
        (
            torch.randn(3, 2),
            torch.randn(6, 3, 3),
            torch.randn(6, 3, 2),
        ),  # Larger batch size for mat1 and mat2
        (
            torch.randn(5, 2),
            torch.randn(4, 5, 5),
            torch.randn(4, 5, 2),
        ),  # Higher dimension tensors for batch
    ),
)
@pytest.mark.parametrize("alpha", [-10.0, -5.0, 0.0, 1.0, 5.0, 10.0])
@pytest.mark.parametrize("beta", [-20.0, -10.0, 0.0, 1.0, 10.0, 20.0])
def test_AddBMM(input, batch1, batch2, alpha, beta):

    # Create an instance of AddBMM
    ch = custom_module.AddBMM()

    # Perform custom AddBMM
    custom_AddBMM = ch.compute_addbmm(input, batch1, batch2, alpha=alpha, beta=beta)

    torch_AddBMM = torch.addbmm(input, batch1, batch2, alpha=alpha, beta=beta)

    # Assert that the custom result matches the expected output
    assert torch.allclose(
        custom_AddBMM, torch_AddBMM, rtol=1e2, atol=1e2
    ), f"custom_AddBMM({input}, {batch1}, {batch2}) != torch_AddBMM({input}, {batch1}, {batch2})"
